### DISCUSSION TITLE
Suggestion de technos 👨‍💻

### DISCUSSION THREAD

**Discussion by maxgfr** [SocialGouv/code-du-travail-numerique#3845] [Technique]
Bonjour à tous, j'ouvre cette discussion concernant des technologies / lib / frameworks qu'on peut implémenter au sein de l'application.

**Comment by maxgfr** [comment]
https://nightwatchjs.org/ <=> Pour les tests e2e

**Comment by maxgfr** [comment]
https://docs.cypress.io/guides/component-testing/introduction#Getting-Started <=> avec cypress on peut faire du component testing

**Reply by maxgfr** [reply]
Le `mount` de Cypress est le `render` de testing-library

**Comment by maxgfr** [comment]
https://storybook.js.org/docs/react/workflows/snapshot-testing <=> avec storybook, en plus de visualiser nos composants, on peut snapshoter le composant de manière unitaire

**Comment by maxgfr** [comment]
https://github.com/testing-library/react-testing-library <=> Pour tester de manière unitaire nos composants

**Reply by maxgfr** [reply]
`testing-library` m'a l'air pas mal du tout pour remplacer `jest` au niveau du test des des composants

**Reply by m-maillot** [reply]
Exemples d'utilisation:
Simple : https://github.com/SocialGouv/archimail/blob/dev/tests/components/Button.test.tsx
Plus avancé : https://github.com/kentcdodds/bookshelf/blob/main/src/__tests__/book-screen.js#L51

Tips provenant d'une discussion mattermost:
L'idée est de se mettre à la place de l'utilisateur, donc ne pas sélectionner par className, data-xxx ou autre. 
Mais plutôt expect(screen.getByRole('button', {name: /add to list/i})).toBeInTheDocument()

Il faut bien comprendre aussi la différence entre les queryBy, getBy et findBy (cf. le cheatsheet React https://testing-library.com/docs/react-testing-library/cheatsheet)

Et pour t'aider à trouver les bonnes queries : https://testing-playground.com/

Encore 2 trucs que j'utilise tout le temps : 
https://testing-library.com/docs/ecosystem-user-event => une meilleure gestion d'évènement que fireEvent
Et le matcher amélioré pour le DOM : https://github.com/testing-library/jest-dom => pour avoir genre toBeInTheDocument

Pas mal de liens intéressant :) A nos claviers ^^

**Comment by maxgfr** [comment]
https://github.com/enzymejs/enzyme <=> Un peu comme `testing-library`

**Comment by maxgfr** [comment]
https://swc.rs/ ; https://esbuild.github.io/ <=> dans le cas où on veut transformer des package avec du typescript

**Comment by maxgfr** [comment]
Au final, on a adopté : 
- cypress pour les tests e2e
- testing/library pour ceux d'inté
- jest pour les unitaires

Et notre bundle est en rust avec swc
